opentelemetry-api/opentelemetry/loader.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+The OpenTelemetry loader module is mainly used internally to load the
+implementation for global objects like :func:`opentelemetry.trace.tracer`.
+
+When loading an implementation, the following algorithm is used:
+
+    1. If the environment variable
+       :samp:`OPENTELEMETRY_PYTHON_IMPLEMENTATION_{getter-name}` (e.g.,
+       ``OPENTELEMETRY_PYTHON_IMPLEMENTATION_TRACER``) is set to an nonempty
+       value, an attempt is made to import a module with that name and call a
+       function ``get_opentelemetry_implementation`` in it.  The function
+       receives the API type that is expected as an argument and should return
+       an instance of it or ``None`` (e.g., the argument is
+       :class:`opentelemetry.trace.Tracer` and the function should return an
+       instance of a :class:`~opentelemetry.trace.Tracer` (probably of a
+       derived type).
+    2. Otherwise, ``OPENTELEMETRY_PYTHON_IMPLEMENTATION_DEFAULT`` is tried
+       instead.
+    3. Otherwise, if a :samp:`set_preferred_{<type>}_implementation` was
+       called, the callback set there is used.
+    4. Otherwise, if :func:`set_preferred_default_implementation` was called,
+       the callback set there is used.
+    5. Otherwise, an attempt is made to import and use the OpenTelemetry SDK.
+    6. Otherwise the default implementation that ships with the API
+       distribution (a fast no-op implementation) is used.
+
+If any of the above steps fails (e.g., a module is loaded but does not define
+the required function or a module name is set but the module fails to load),
+the search immediatelly skips to the last step.
+"""
+
+from typing import Type, TypeVar, Optional, Callable, Dict
+import importlib
+import os
+import sys
+
+_T = TypeVar('_T')
+
+_UntrustedImplFactory = Callable[[Type[_T]], object]
+
+#ImplementationFactory = Callable[[Type[_T]], _T]
```

Comment:
Do you mean to leave this in? And why bother calling the type "untrusted"?

It's also not clear why we need `object` instead of the generic type here, but I'm willing to believe that we do. FWIW I don't think we should use type annotations outside of the overridable API components unless we have reason to believe that that bit of code in particular would benefit from type checks.

Reply:
It's called untrusted and returns an object because we don't trust it to return the type we expect, since it is set by code that is out of our control. Using `object` forces us to check the type (as `object` has basically no methods, is not convertible to anything, etc.).

`ImplementationFactory` was meant for the other side, the `set_preferred_tracer_implementation*_impl` functions. But due to bug python/mypy#7092, I had to repeat the definition in the setter.

I'd actually prefer to leave it here as documentation, but I should add a comment explaining this.

Reply:
I added comments, is that OK?

Reply:
> FWIW I don't think we should use type annotations outside of the overridable API components

Presently we run mypy with `disallow_untyped_defs = True` so it wouldn't even pass travis. And personally I like static typing.

Reply:
> And personally I like static typing

No accounting for taste :P